From 357677aed6f80571c3cf7d5a13e61abb5428d2f7 Mon Sep 17 00:00:00 2001
From: Wang Wensheng <wangwensheng4@huawei.com>
Date: Tue, 18 Apr 2023 09:10:37 +0000
Subject: [PATCH 39/53] [Huawei] mm/sharepool: Delete useless error recovery
 code in sp_group_add_task

Offering: HULK
hulk inclusion
category: feature
bugzilla: N/A

--------------------------------

Since we ignore the error code when we do mmap for the existing spa,
no error would occur after sp_map_spa_to_mm(). Delete the dead code.

Also change the return value of group_add_task(). since no one need
a sp_group_node from group_add_task().

Signed-off-by: Wang Wensheng <wangwensheng4@huawei.com>
---
 mm/share_pool.c | 29 ++++++++++-------------------
 1 file changed, 10 insertions(+), 19 deletions(-)

diff --git a/mm/share_pool.c b/mm/share_pool.c
index a3b687057983..571adf4e7b3d 100644
--- a/mm/share_pool.c
+++ b/mm/share_pool.c
@@ -486,14 +486,13 @@ static int sp_mapping_group_setup(struct mm_struct *mm, struct sp_group *spg)
 
 static struct sp_group *sp_group_create(int spg_id);
 static void free_sp_group_locked(struct sp_group *spg);
-static struct sp_group_node *group_add_task(struct mm_struct *mm, struct sp_group *spg,
+static int group_add_task(struct mm_struct *mm, struct sp_group *spg,
 					    unsigned long prot);
 static int init_local_group(struct mm_struct *mm)
 {
 	int ret;
 	struct sp_group *spg;
 	struct sp_mapping *spm;
-	struct sp_group_node *spg_node;
 	struct sp_group_master *master = mm->sp_group_master;
 
 	spg = sp_group_create(SPG_ID_LOCAL);
@@ -510,12 +509,10 @@ static int init_local_group(struct mm_struct *mm)
 	sp_mapping_attach(master->local, sp_mapping_normal);
 	sp_mapping_attach(master->local, sp_mapping_ro);
 
-	spg_node = group_add_task(mm, spg, PROT_READ | PROT_WRITE);
-	if (IS_ERR(spg_node)) {
+	ret = group_add_task(mm, spg, PROT_READ | PROT_WRITE);
+	if (ret)
 		/* The spm would be released while destroying the spg */
-		ret = PTR_ERR(spg_node);
 		goto free_spg;
-	}
 	mmget(mm);
 
 	return 0;
@@ -1265,23 +1262,23 @@ static void free_spg_node(struct mm_struct *mm, struct sp_group *spg,
 }
 
 /* the caller must hold sp_global_sem and down_write(&spg->rw_lock) in order */
-static struct sp_group_node *group_add_task(struct mm_struct *mm, struct sp_group *spg,
-					    unsigned long prot)
+static int group_add_task(struct mm_struct *mm, struct sp_group *spg,
+			  unsigned long prot)
 {
 	struct sp_group_node *node;
 	int ret;
 
 	node = create_spg_node(mm, prot, spg);
 	if (IS_ERR(node))
-		return node;
+		return PTR_ERR(node);
 
 	ret = insert_spg_node(spg, node);
 	if (unlikely(ret)) {
 		free_spg_node(mm, spg, node);
-		return ERR_PTR(ret);
+		return ret;
 	}
 
-	return node;
+	return 0;
 }
 
 /**
@@ -1309,7 +1306,6 @@ int mg_sp_group_add_task(int tgid, unsigned long prot, int spg_id)
 	struct task_struct *tsk;
 	struct mm_struct *mm;
 	struct sp_group *spg;
-	struct sp_group_node *node = NULL;
 	int ret = 0;
 	struct sp_area *spa;
 
@@ -1390,10 +1386,9 @@ int mg_sp_group_add_task(int tgid, unsigned long prot, int spg_id)
 		goto out_drop_group;
 	}
 
-	node = group_add_task(mm, spg, prot);
-	if (unlikely(IS_ERR(node))) {
+	ret = group_add_task(mm, spg, prot);
+	if (ret) {
 		up_write(&spg->rw_lock);
-		ret = PTR_ERR(node);
 		goto out_drop_group;
 	}
 
@@ -1411,12 +1406,8 @@ int mg_sp_group_add_task(int tgid, unsigned long prot, int spg_id)
 		}
 	}
 
-	if (unlikely(ret))
-		delete_spg_node(spg, node);
 	up_write(&spg->rw_lock);
 
-	if (unlikely(ret))
-		free_spg_node(mm, spg, node);
 	/*
 	 * to simplify design, we don't release the resource of
 	 * group_master and proc_stat, they will be freed when
-- 
2.17.1

